src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -0,0 +1,448 @@
+                case ',': return RegexKind.CommaToken;
+                case ':': return RegexKind.ColonToken;
+                case '=': return RegexKind.EqualsToken;
+                case '!': return RegexKind.ExclamationToken;
+                case '<': return RegexKind.LessThanToken;
+                case '>': return RegexKind.GreaterThanToken;
+                case '-': return RegexKind.MinusToken;
+                case '\'': return RegexKind.SingleQuoteToken;
+                default: return RegexKind.TextToken;
+            }
+        }
+
+        private ImmutableArray<RegexTrivia> ScanLeadingTrivia(bool allowTrivia, RegexOptions options)
+        {
+            if (!allowTrivia)
+            {
+                return ImmutableArray<RegexTrivia>.Empty;
+            }
+
+            var result = ArrayBuilder<RegexTrivia>.GetInstance();
+
+            var start = Position;
+
+            while (Position < Text.Length)
+            {
+                var comment = ScanComment(options);
+                if (comment != null)
+                {
+                    result.Add(comment.Value);
+                    continue;
+                }
+
+                var whitespace = ScanWhitespace(options);
+                if (whitespace != null)
+                {
+                    result.Add(whitespace.Value);
+                    continue;
+                }
+
+                break;
+            }
+
+            return result.ToImmutableAndFree();
+        }
+
+        public RegexTrivia? ScanComment(RegexOptions options)
+        {
+            if (Position < Text.Length)
+            {
+                if (HasOption(options, RegexOptions.IgnorePatternWhitespace))
+                {
+                    if (Text[Position] == '#')
+                    {
+                        var start = Position;
+
+                        // Note: \n is the only newline the native regex parser looks for.
+                        while (Position < Text.Length &&
+                               Text[Position] != '\n')
+                        {
+                            Position++;
+                        }
+
+                        return CreateTrivia(RegexKind.CommentTrivia, GetSubPattern(start, Position));
+                    }
+                }
+
+                if (IsAt("(?#"))
+                {
+                    var start = Position;
+                    while (Position < Text.Length &&
+                           Text[Position] != ')')
+                    {
+                        Position++;
+                    }
+
+                    if (Position == Text.Length)
+                    {
+                        var diagnostics = ImmutableArray.Create(new EmbeddedDiagnostic(
+                            WorkspacesResources.Unterminated_regex_comment,
+                            GetTextSpan(start, Position)));
+                        return CreateTrivia(RegexKind.CommentTrivia, GetSubPattern(start, Position), diagnostics);
```

Comment:
Is there a case where trivia holds more than one diagnostic? If not, consider just holding one (or `null`).

Reply:
I don't know if this was addressed. The question is why store an array of diagnostics? It seems we only need zero or one.

---
In reply to: [178448065](https://github.com/dotnet/roslyn/pull/23984#discussion_r178448065) [](ancestors = 178448065)